### PR TITLE
Issue 174 test unlock

### PR DESCRIPTION
**Pull request by spgarbet**

### CODE REVIEW

File: R/unlockREDCap.R [hunk 1]
```diff
@@ -15,15 +14,120 @@
...
+  
+  return(dest)
+}
+ 
+  #############################################################################
+ ## unlock keyring
+##
+.unlockKeyring <- function(keyring, passwordFUN)
+{
+  state <- keyring_list()
```
Comment (couthcommander) [inline R/unlockREDCap.R:77]:
```suggestion
  state <- keyring::keyring_list()
```

File: R/unlockREDCap.R [hunk 2]
```diff
@@ -15,15 +14,120 @@
...
+      if(is.null(password) || password == '') stop(paste0("User aborted keyring '",keyring, "' unlock."))
+      
+      tryCatch(
+        {
+          keyring_unlock(keyring, password)
+          Sys.setenv(REDCAPAPI_PW=password)
+          # Hacked work around for RStudio starting new session for everything
+          if(requireNamespace("rstudioapi", quietly = TRUE) &&
+            rstudioapi::isAvailable(child_ok=TRUE))
+            rstudioapi::sendToConsole(paste0("Sys.setenv(REDCAPAPI_PW='", password, "')"), execute = TRUE, echo=FALSE)
```
Comment (couthcommander) [inline R/unlockREDCap.R:100]:
```suggestion
            rstudioapi::sendToConsole(sprintf("Sys.setenv(REDCAPAPI_PW='%s')", password), execute = TRUE, echo=FALSE)
```

Comment (couthcommander) [reply R/unlockREDCap.R:100]: very minor change; I prefer `sprintf` when more than one comma in `paste` command

File: R/unlockREDCap.R [hunk 3]
```diff
@@ -132,124 +237,53 @@ unlockREDCap    <- function(connections,
...
-  state <- state[state$keyring==keyring,]
-  
-  # If so, does it exist?
-  if(nrow(state) == 1)
+  # Open Connections
+  dest <- lapply(seq_along(connections), function(i)
   {
-    # Is it locked
-    if(state$locked)
+    stored <- connections[i] %in% key_list("redcapAPI", keyring)[,2]
```
Comment (couthcommander) [inline R/unlockREDCap.R:251]:
```suggestion
    stored <- connections[i] %in% keyring::key_list("redcapAPI", keyring)[,2]
```

File: R/unlockREDCap.R [hunk 4]
```diff
@@ -132,124 +237,53 @@ unlockREDCap    <- function(connections,
...
-    # Check again if it's set properly
-    if(!.key_saved(apiKeyStore, connections[i]))
-    {
-      key <- passwordFUN(paste("Please enter RedCap API_KEY for", connections[i]))
-      
-      if(is.null(key) || key == '') stop(paste("No Key Entered for", connections[i]))
-      
-      apiKeyStore[[connections[i]]] <- key
-      if(!is.null(keyring))
+        key_delete("redcapAPI", unname(connections[i]), keyring)
```
[the remark this answers is not in the document]
Comment (couthcommander) [reply R/unlockREDCap.R:269]: `key_delete` is not imported in NAMESPACE

File: R/unlockREDCap.R [hunk 5]
```diff
@@ -132,124 +237,53 @@ unlockREDCap    <- function(connections,
...
+        api_key <- passwordFUN(paste0(
+          "Invalid API_KEY for '", connections[i],
+          "' in keyring '", keyring,
+          "'. Possible causes include: mistyped, renewed, or revoked.",
+          " Please enter a new key or cancel to abort."))
+        if(is.null(api_key) || api_key == '') stop("unlockREDCap aborted")
+      } else if(!stored)
       {
-        keyring::key_set_with_value("redcapAPI", username=connections[[i]], password=apiKeyStore[[connections[i]]], keyring=keyring)
+        key_set_with_value( service="redcapAPI",
```
Comment (couthcommander) [inline R/unlockREDCap.R:278]:
```suggestion
        keyring::key_set_with_value( service="redcapAPI",
```

File: R/unlockREDCap.R [hunk 6]
```diff
@@ -15,15 +14,120 @@
...
+        }
+      )
+    }
+  } else # Keyring does not exist => Create
+  {
+    password <- passwordFUN(paste0("Creating keyring. Enter NEW password for the keyring '",
+                                   keyring, "'."))
+    if(is.null(password) || password == '') stop(paste0("User cancelled creation of keyring '", keyring, "'."))
+
+    keyring_create(keyring, password)
```
Comment (couthcommander) [inline R/unlockREDCap.R:125]:
```suggestion
    keyring::keyring_create(keyring, password)
```

File: R/unlockREDCap.R [hunk 7]
```diff
@@ -15,15 +14,120 @@
...
+    while(locked)
+    {
+      password <- Sys.getenv("REDCAPAPI_PW")
+      stored   <- !is.null(password) && password != ''
+      if(!stored) password <- passwordFUN(msg)
+      if(is.null(password) || password == '') stop(paste0("User aborted keyring '",keyring, "' unlock."))
+      
+      tryCatch(
+        {
+          keyring_unlock(keyring, password)
```
Comment (couthcommander) [inline R/unlockREDCap.R:95]:
```suggestion
          keyring::keyring_unlock(keyring, password)
```

File: R/unlockREDCap.R (hunk 1 above)
Comment (spgarbet) [reply R/unlockREDCap.R:77]: I changed this to the non-namespaced version to work with the stubbed tests.

File: R/unlockREDCap.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (spgarbet) [reply R/unlockREDCap.R:284]: I fixed this

File: R/unlockREDCap.R (hunk 2 above)
Comment (spgarbet) [reply R/unlockREDCap.R]: Done.

**Review by couthcommander — APPROVED** [review]
Looks fine; I recommend PKG::FUN syntax, though it is creating problems with mocked tests.

**Pull request merged**